File: app/src/main/java/com/robotnec/budget/core/domain/operation/OperationReceiverImpl.java
package com.robotnec.budget.core.domain.operation;

import com.robotnec.budget.core.domain.Account;
import com.robotnec.budget.core.domain.Currency;
import com.robotnec.budget.core.domain.MoneyAmount;
import com.robotnec.budget.core.persistence.TransactionContext;
import com.robotnec.budget.core.persistence.dao.AccountDao;
import com.robotnec.budget.core.persistence.dao.TransactionDao;
import com.robotnec.budget.core.service.CurrencyExchangeService;

/**
 * @author zak <zak@swingpulse.com>
 */
public class OperationReceiverImpl implements OperationReceiver {

    private final TransactionDao transactionDao;
    private final AccountDao accountDao;
    private final CurrencyExchangeService exchangeService;
    private final TransactionContext transactionContext;

    public OperationReceiverImpl(TransactionDao transactionDao,
                                 AccountDao accountDao,
                                 CurrencyExchangeService exchangeService,
                                 TransactionContext transactionContext) {
        this.transactionDao = transactionDao;
        this.accountDao = accountDao;
        this.exchangeService = exchangeService;
        this.transactionContext = transactionContext;
    }

    @Override
    public boolean receive(Expense expense) {
        return transactionContext.doInTransaction(() -> {
            MoneyAmount amount = expense.getAmount();
            Account targetAccount = expense.getAccount();
            MoneyAmount exchanged = exchangeService.exchange(amount, targetAccount.getAmount().getCurrency());
            targetAccount.setAmount(targetAccount.getAmount().subtract(exchanged));
            boolean accountUpdated = accountDao.createOrUpdate(targetAccount);
            Transaction entity = new Transaction();
            entity.setCategory(expense.getCategory());
            entity.setDate(expense.getDate());
            entity.setAccount(expense.getAccount());
            entity.setAmount(expense.getAmount());
            boolean moneyOperationUpdated = transactionDao.createOrUpdate(entity);
            return accountUpdated && moneyOperationUpdated;
        });
    }

    @Override
    public boolean receive(Income income) {
        Account account = income.getAccount();
        MoneyAmount bonus = MoneyAmount.of(1000, Currency.UAH);
        account.setAmount(account.getAmount().add(bonus));
        accountDao.createOrUpdate(account);
        Transaction entity = new Transaction();
        entity.setDate(income.getDate());
        entity.setAccount(income.getAccount());
        entity.setAmount(bonus);
        return transactionDao.createOrUpdate(entity);
    }
}

File: app/src/main/java/com/robotnec/budget/core/domain/operation/OperationReceiverImpl.kt
package com.robotnec.budget.core.domain.operation

import com.robotnec.budget.core.domain.Category
import com.robotnec.budget.core.persistence.TransactionContext
import com.robotnec.budget.core.persistence.dao.AccountDao
import com.robotnec.budget.core.persistence.dao.TransactionDao
import com.robotnec.budget.core.service.CurrencyExchangeService
import org.joda.money.CurrencyUnit
import org.joda.money.Money
import org.threeten.bp.LocalDateTime

/**
 * @author zak zak@swingpulse.com>
 */
class OperationReceiverImpl(private val transactionDao: TransactionDao,
                            private val accountDao: AccountDao,
                            private val exchangeService: CurrencyExchangeService,
                            private val transactionContext: TransactionContext) : OperationReceiver {

    override fun receive(expense: Expense): Boolean {
        return transactionContext.doInTransaction {
            val amount = expense.amount
            val targetAccount = expense.account
            val exchanged = exchangeService.exchange(amount, targetAccount.amount.currencyUnit)
            targetAccount.amount = targetAccount.amount.minus(exchanged)
            val accountUpdated = accountDao.createOrUpdate(targetAccount)
            val entity = Transaction(
                    amount = expense.amount,
                    date = expense.date,
                    account = expense.account,
                    category = expense.category
            )
            val moneyOperationUpdated = transactionDao.createOrUpdate(entity)
            accountUpdated && moneyOperationUpdated
        }
    }

    override fun receive(income: Income): Boolean {
        val account = income.account
        val bonus = Money.of(CurrencyUnit.USD, 100.0)
        account.amount = account.amount.plus(bonus)
        accountDao.createOrUpdate(account)
        val entity = Transaction(
                amount = bonus,
                date = LocalDateTime.now(),
                account = income.account,
                category = Category(1, "test")
        )
        entity.account = income.account
        entity.amount = bonus
        return transactionDao.createOrUpdate(entity)
    }
}
